vesting/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+	fn locked(who: &T::AccountId) -> Option<(BalanceOf<T>, T::BlockNumber)> {
+		let now = <frame_system::Module<T>>::block_number();
+		Self::vesting_schedules(who).iter().fold(None, |acc, s| {
+			let locked_amount = s.locked_amount(now);
+			if locked_amount.is_zero() {
+				return acc;
+			}
+
+			let s_end = s.end().expect("ensured not overflow while adding; qed");
+			if let Some((amount, until)) = acc {
+				Some((amount + locked_amount, until.max(s_end)))
+			} else {
+				Some((locked_amount, s_end))
+			}
+		})
+	}
+
+	fn do_add_vesting_schedule(
+		from: &T::AccountId,
+		to: &T::AccountId,
+		schedule: VestingScheduleOf<T>,
+	) -> DispatchResult {
+		Self::ensure_lockable(to)?;
+
+		let (schedule_amount, schedule_end) = Self::ensure_valid_vesting_schedule(&schedule)?;
+		let (mut total_amount, mut until) = (schedule_amount, schedule_end);
+		if let Some((curr_amount, curr_until)) = Self::locked(to) {
+			total_amount = curr_amount
+				.checked_add(&schedule_amount.into())
+				.ok_or(Error::<T>::NumOverflow)?;
+			until = until.max(curr_until);
+		}
+
+		T::Currency::transfer(from, to, schedule_amount, ExistenceRequirement::AllowDeath)?;
+		T::Currency::set_lock(VESTING_LOCK_ID, to, total_amount, until, WithdrawReasons::all());
+		<VestingSchedules<T>>::mutate(to, |v| (*v).push(schedule));
+
+		Ok(())
+	}
+
+	fn do_update_vesting_schedules(who: &T::AccountId, schedules: Vec<VestingScheduleOf<T>>) -> DispatchResult {
+		Self::ensure_lockable(who)?;
+
+		let (total_amount, until) = schedules
+			.iter()
+			.try_fold::<_, _, Result<(BalanceOf<T>, T::BlockNumber), Error<T>>>(
+				(Zero::zero(), Zero::zero()),
+				|(acc_amount, acc_end), schedule| {
+					let (amount, end) = Self::ensure_valid_vesting_schedule(schedule)?;
+					Ok((acc_amount + amount, acc_end.max(end)))
+				},
+			)?;
+		ensure!(
+			T::Currency::free_balance(who) >= total_amount,
+			Error::<T>::InsufficientBalanceToLock,
+		);
+
+		T::Currency::set_lock(VESTING_LOCK_ID, who, total_amount, until, WithdrawReasons::all());
+		<VestingSchedules<T>>::insert(who, schedules);
+
+		Ok(())
+	}
+
+	/// Returns `Ok((amount, end))` if valid schedule, or error.
+	fn ensure_valid_vesting_schedule(
+		schedule: &VestingScheduleOf<T>,
+	) -> Result<(BalanceOf<T>, T::BlockNumber), Error<T>> {
+		ensure!(!schedule.period.is_zero(), Error::<T>::ZeroVestingPeriod);
+		ensure!(!schedule.period_count.is_zero(), Error::<T>::ZeroVestingPeriodCount);
+
+		let amount = schedule.total_amount().ok_or(Error::<T>::NumOverflow)?;
+		let schedule_end = schedule.end().ok_or(Error::<T>::NumOverflow)?;
+		Ok((amount, schedule_end))
+	}
+
+	/// Ensure no other types of locks except `VESTING_LOCK_ID`.
+	fn ensure_lockable(who: &T::AccountId) -> DispatchResult {
+		// FIXME: use locks query in `LockableCurrency` when it's ready
+		// https://github.com/paritytech/substrate/issues/4655
+
+		// FIXME: remove `do_claim` workaround after issue #4655 fixed
```

Comment:
double FIXME
can you add a Github issue about this to make it easier to track

Reply:
Will do. I was planning to add a issue after this PR merged.